src/firefly/js/ui/panel/TabPanel.jsx

```diff
@@ -156,71 +159,74 @@ Tabs.propTypes= {
     borderless: PropTypes.bool
 };
 
-Tabs.defaultProps= {
+StatelessTabs.defaultProps= {
     defaultSelected: 0,
     useFlex: false,
     resizable: false,
     borderless: false
 };
 
 
-export class Tab extends PureComponent {
-    constructor(props) {
-        super(props);
-    }
+/**
+ * TabPanel with internal state
+ * State will be lost after unload.
+ */
+export const Tabs = React.memo( (props) => {
+    const {defaultSelected=0, onTabSelect} = props;
+
+    const [selectedIdx, setSelectedIdx] = useState(defaultSelected);
 
-    componentDidMount() {
-        const {selected, onSelect, id, name} = this.props;
-        if (selected) {
-            onSelect(id, name);
+    const onSelect = useCallback( (index,id,name) => {
+        if (index !== selectedIdx) {
+            setSelectedIdx(index);
+            onTabSelect && onTabSelect(index,id,name);
         }
-    }
+    });
+
+    return (<StatelessTabs {...props} defaultSelected={selectedIdx} onTabSelect={onSelect} />);
+});
+
+Tabs.propTypes = StatelessTabs.propTypes;
+Tabs.defaultProps = StatelessTabs.defaultProps;
+
 
-    render () {
-        const {name, label, selected, onSelect, removable, onTabRemove, id, maxTitleWidth} = this.props;
+/*----------------------------------------------------------------------------------------------*/
+/**
+ * TabPanel with ComponentCntlr supported state
+ * Selected state is stored as <componentKey>.selectedIdx
+ */
+export const StatefulTabs = React.memo( (props) => {
+    const {children=[], defaultSelected=0, onTabSelect, componentKey} = props;
+
+    const [selectedIdx = defaultSelected] = useStoreConnector( () => {
+        return get(getComponentState(componentKey), 'selectedIdx');
+    });
+
+    const onSelect = useCallback( (index,id,name) => {
+        dispatchComponentStateChange(componentKey, {selectedIdx: index});
+        onTabSelect && onTabSelect(index,id,name);
+    }, []);
 
-        let tabClassName = 'TabPanel__Tab' ;
-        if (selected) {
-            tabClassName += ' TabPanel__Tab--selected';
+    useEffect( ()=> {
+        if (selectedIdx >= children.length) {
+            // selectedIdx is greater than the number of tabs.. update store's state
+            selectedIdx = children.length-1;
```

Comment:
this is a constant being reassigned.

Reply:
thanks..  eslint didn't catch it.  did the error show up for you?

Reply:
it show up through intellij not ESlint.  We probably should have some eslint rule.